### ISSUE TITLE
Relative image paths don't resolve without file:// prefix

### ISSUE THREAD

**Issue by dejastew2** [pbek/QOwnNotes#803]
I've seen in other reported issues that QON uses relative image links prefixed by `file://` to differentiate it from `note://`. Is it possible to make the default behavior of a relative image link with _no_ prefix to be a relative link to an image? This appears to match the expected behavior of other Markdown editors I've experimented with (notably neutriNote and Epsilon Notes on Android); unfortunately this difference means that I can either have inline images appear in QON, _or_ all other Markdown editors. 

Example in QON:

![qon_relative_path_issue](https://user-images.githubusercontent.com/134419/33521208-98c0f8ee-d780-11e7-8f61-256f900525d6.png)

Example in neutriNote:

![screenshot_20171202-163422](https://user-images.githubusercontent.com/134419/33521213-a83d54de-d780-11e7-9ed7-5a07558004b0.jpg)

If it matters, I am running QON on Windows 10 with version 17.12.0, and my image links are `![QON relative link](file://media/bmw_m6.jpg)` and `![Removing the file://](media/bmw_m6.jpg)`.

**Comment by dejastew2** [comment]
I looked very closely at that issue, but it seemed to be a request for path independence (a .md file in a subfolder being able to use `media/file.jpg` instead of `../media/file.jpg`, for example). I am completely fine with relative paths being exact (`../media/file.jpg`), I just want to not have to prepend `file://` to the relative path (to increase portability). This is similar to what @obmk requested in #266, but that issue has developed quite a bit of scope creep, and I thought a second, more targeted issue was worth opening so we could discuss that specific feature request.

**Comment by pbek** [comment]
#266 is all about getting rid of the url-scheme and all attempts to do that for the highlighter and the preview introduced a truckload of troubles. I think that is what you were requesting, am I right?

**Comment by dejastew2** [comment]
Yes, I was hoping to get relative links without the `file://` to work in the preview window. If #266 is a sufficient placeholder for those efforts, then that's fine. 

I reread #266 and discovered it actually contains the answer to my real question, which is "how do I get an inline image to appear in both QON and in an Android Markdown app". @Maboroshy mentioned that Epsilon Notes (which I've been using on my Android phone) contains a _compatibility preprocessor_ which allows the user to enter a custom regex that will run when a Markdown file is about to be rendered in its preview window. This allows you to do things like replace `![$1](file://$2)` with `![$1]($2)` without actually modifying the saved Markdown note file. For anyone who wants to do the same thing and comes across this issue, this is the regex string I used to remove all `file://` prefixes from links:

    \[([^\]]*?)\]\(file://([^\)]*?)\)
    [$1]($2)

@pbek, thanks for the support and all your efforts in developing QON!

**Comment by pbek** [comment]
Thank you for your kind words!

**Comment by Maboroshy** [comment]
Simply removing `file://` without providing absolute path to media or attachments can go wrong with notes in subfolders. 
It's the reason I explicitly used absolute paths in my example.

**Comment by dejastew2** [comment]
Thanks for clarifying; I hadn't tested subfolders yet. For those who use subfolders, and use Dropbox as a way to sync their notes, the updated compatibility preprocessor regex I'm using now looks like this (replace u1234567 with your user ID, which you can find by browsing to the path on your Android device):

    \[([^\]]*?)\]\(file://([^\)]*?)\)
    [$1](/storage/emulated/0/Android/data/com.dropbox.android/files/u1234567/scratch/Notes/$2)

**Comment by Offerel** [comment]
Sorry if I bring this subject out again. Currently I am trying to find a way to get Markdown notes on the different platforms to work together. On the webserver I don't use an Owncloud/Nextcloud, but a flat directory structure. Notes are stored in a single folder. If these notes contain images, they are located in a subfolder or encoded as base64 in the note itself. The later is great for portability, but some editors are slowed down enormously with base64 encoded images, so I try to avoid it as much as possible. I have now teached all editors that the contained images are located in the subfolder "media". Thus I can access pictures with the relative path "media/image12345. jpg" from the note. This works online, with various JavaScript editors, but also offline with programs like MarkdownPad, Texts and much more. Even under Android all editors are able to display the images correctly. On Windows I would like to use QOwnNotes now, because it is the only program I know so far, which gives me a list of notes from folder "xyz", all the others only work with single notes or I have to open them manually. 

The approach, which places inserted images with a new name in the media subfolder, fits so far to my workflow shown above but unfortunately the protocol "file: //" is inserted in front of each image. All other editors can't handle this. However, if I remove the protocol manually in QOwnNotes, the images are again not displayed correctly in QOwnNotes. Isn't there an option that can display QOwnNotes without "file: //" images, just like any other program?

I have also tried using scripting. The script absolute-media-links.qml but changed there `return mediaMarkdownText. replace (new RegExp ("file: //","g"),"");` but also this does not remove "file: //" from me. I can't see any effect from the script at all. In the Markdown will continue to insert !Added[graphic (3)] (file: //media/278. jpeg).

I suppose I just don't understand the sense of scripting or it doesn't really work the way I want it to. If the preview of QOwnNotes can only display images with "file: //", I'd like to have a script, which is possible with!(media/278. jpeg) is inserted into the markdown, but in the preview it will be shown with!Graphic (3)] (file: //media/278. jpeg) will be replaced. But to be honest, it's a mystery how I have to put this into practice. Somebody might be able to help me out here.

**Comment by Offerel** [comment]
OK, think i'm on the right way. I have changed the qml now to:
```
function insertMediaHook(fileName, mediaMarkdownText) {
        return mediaMarkdownText.replace(new RegExp("file://", "g"), "");
    }
```

This inserts now the correct syntax in to the Markdown file. Now the preview does not work as expected. I have added in to the same qml file now a additional function:

```
function noteToMarkdownHtmlHook(note, html) {
        var path = script.currentNoteFolderPath();
		html = html.replace("src=\"media", "src=\"file://" + path + "/media");
		return html;
	}
```

It seems to work and changes the path in the preview to a absolute path. However it takes really long and QOwnNotes freezes for some seconds before it comes back. But the image isnt displayed. However, if i export the html, and test this file in Internet Explorer, it displays the picture. IE displays also a exported HTML without the second function, with the relative path.

So in my case, inserting is fine now, but what can i do to make the preview work?

**Comment by Maboroshy** [comment]
1. Absolute path to image works without `file://`.
2. I've never tried it, but you can try using `noteToMarkdownHtmlHook` scripting hook to modify the path to image in html QON generates from markdown. If you have path without `file://`in your note you'll get `<img src="/media/image.jpg" alt=""/>` at html, which is broken and won't render. You can try using `noteToMarkdownHtmlHook` to turn it to an absolute path or something else that QON would render.

**Comment by Offerel** [comment]
When i insert the above regex
 ```
\[([^\]]*?)\]\(file://([^\)]*?)\)
[$1]($2)
``` 
into Epsion Notes and you get a error, i have now realized, that this needs to be on 2 lines. There is a linebreak just before `[$1]($2)`. Copy and paste removed this on my device.

**Comment by pbek** [comment]
Could you please head over to #266 to test and comment on the new features.

**Comment by dejastew2** [comment]
Updated #266, but to save people a click: this fixes the compatibility issue I was seeing previously when syncing with other Markdown editors. No more preprocessors needed -- thanks @pbek!

**Comment by pbek** [comment]
Thanks a lot for testing!
